kubespawner/spawner.py

```diff
@@ -2633,3 +2633,27 @@ async def _ensure_namespace(self):
                 # It's fine if it already exists
                 self.log.exception("Failed to create namespace %s", self.namespace)
                 raise
+
+    async def delete_forever(self):
+        """Called when a user is deleted.
+
+        This can do things like request removal of resources such as persistent storage.
+        Only called on stopped spawners, and is likely the last action ever taken for the user.
+
+        Will only be called once on the user's default Spawner.
+        """
+        try:
+            self.api.read_namespaced_persistent_volume_claim_with_http_info(
+                self.pvc_name, self.namespace
+            )
+        except ApiException as e:
+            if e.status == 404:
+                self.log.warning(
+                    "Could not delete %s. This PVC does not exist.", self.pvc_name
+                )
+            else:
+                raise
+        else:
+            self.api.delete_namespaced_persistent_volume_claim(
+                self.pvc_name, self.namespace
+            )
```

Comment:
Thanks for your work on this @nsshah1288! I think this is a feature to support now that it is part of JupyterHub 1.4.0+, which btw can be worth mentioning in the docstring.

@nsshah1288 can you elaborate on why this deletion logic make sense to structure in try/except/else and what the difference between `read_namespaced_persistent_volume_claim_with_http_info` and without `http_info` is?

Reply:
Thanks @consideRatio ! I have updated the docstring and also removed `_with_http_info` from the first API call. `read_namespaced_persistent_volume_claim` actually calls `read_namespaced_persistent_volume_claim_with_http_info`  in `CoreV1Api`, and the  `_with_http_info` I believe just indicates that the method is making an HTTP request to check if the specified PVC exists. I have not yet tested this change, but I will update here when I do.

I was thinking that when `delete_forever` is called, I first check if the specified PVC exists. If it doesn't exist, I throw an exception, and only if the PVC exists do I execute the `delete_namespaced_persistent_volume_claim`, hence the try/except/else. Does that sound OK?

Reply:
hi @consideRatio , I tested this change and it works well. When clicking 'delete user', it deletes the PVC and PV of the user if it exists. (tested with sqlite and 'dynamic' storage setting in values.yaml). Deleting users that don't have existing PVCs or PVs also works fine.

Reply:
Thanks for verifying that it worked!

With regards to "_with_http_info" or not, I suggest we align with the method for deleting other resources with kubespawner. If we learn that it can make sense to use `_with_http_info` or not we update all our calls in kubespawner.

Reply:
Hi @consideRatio, could you please let me know if anything else is required for this PR? I tried to resolve your comments and also wrote a test case. Any feedback would be great, thanks!

Reply:
Hello @consideRatio , did you had some time to review this change? It would be amazing to have it in the official version. At the moment i'm using a modified version of JH with those changes and works like a charm in Openshift.